src/codegen/encoding/mod.rs

```diff
@@ -10,27 +10,520 @@ use crate::codegen::encoding::scale_encoding::ScaleEncoding;
+                    no_padding_size,
+                ))
+            }
+        } else {
+            None
+        };
+
+        let qty = struct_ty.definition(ns).fields.len();
+        let first_ty = struct_ty.definition(ns).fields[0].ty.clone();
+        let loaded = load_struct_member(first_ty, expr.clone(), 0);
+
+        let mut advance = self.encode(&loaded, buffer, &offset, arg_no, ns, vartab, cfg);
+        let mut runtime_size = advance.clone();
+        for i in 1..qty {
+            let ith_type = struct_ty.definition(ns).fields[i].ty.clone();
+            offset = Expression::Add(
+                Codegen,
+                Uint(32),
+                false,
+                offset.clone().into(),
+                advance.into(),
+            );
+            let loaded = load_struct_member(ith_type.clone(), expr.clone(), i);
+            // After fetching the struct member, we can encode it
+            advance = self.encode(&loaded, buffer, &offset, arg_no, ns, vartab, cfg);
+            runtime_size = Expression::Add(
+                Codegen,
+                Uint(32),
+                false,
+                runtime_size.into(),
+                advance.clone().into(),
+            );
+        }
+
+        size.unwrap_or(runtime_size)
+    }
+
+    /// Encode `expr` into `buffer` as an array.
+    fn encode_array(
+        &mut self,
+        array: &Expression,
+        array_ty: &Type,
+        elem_ty: &Type,
+        dims: &[ArrayLength],
+        arg_no: usize,
+        buffer: &Expression,
+        offset: &Expression,
+        ns: &Namespace,
+        vartab: &mut Vartable,
+        cfg: &mut ControlFlowGraph,
+    ) -> Expression {
+        assert!(!dims.is_empty());
+
+        if allow_direct_copy(array_ty, elem_ty, dims, ns) {
+            // Calculate number of elements
+            let (bytes_size, offset, size_length) =
+                if matches!(dims.last(), Some(&ArrayLength::Fixed(_))) {
+                    let elem_no = calculate_direct_copy_bytes_size(dims, elem_ty, ns);
+                    (
+                        Expression::NumberLiteral(Codegen, Uint(32), elem_no),
+                        offset.clone(),
+                        None,
+                    )
+                } else {
+                    let value = array_length(array, vartab, cfg);
+
+                    let (new_offset, size_length) = if self.is_packed() {
+                        (offset.clone(), None)
+                    } else {
+                        let encoded_size = self.encode_size(&value, buffer, offset, vartab, cfg);
+                        let new_offset = increment_by(offset.clone(), encoded_size.clone());
+                        (new_offset, Some(encoded_size))
+                    };
+
+                    if let Expression::Variable(_, _, size_temp) = value {
+                        let size = calculate_array_bytes_size(size_temp, elem_ty, ns);
+                        (size, new_offset, size_length)
```

Comment:
As you already have an `Expression::Variable` here, what do you think of refactoring `calculate_array_bytes_size` to receive an `Expression` instead of the variable number?

Reply:
Thanks for checking, I'm gonna leave it as it is then.